src/lib/mina_generators/parties_generators.ml

```diff
@@ -745,92 +731,58 @@ let gen_party_body_components (type a b c d) ?(update = None) ?account_id
-            { account_with_gen_pk with
-              public_key = available_pk
-            ; token_id = Token_id.default
-            }
+            Account.create account_id (Currency.Balance.of_int 0)
           in
           let account =
             if zkapp_account then
               { account_with_pk with
                 zkapp =
-                  (let vk =
-                     match vk with
-                     | None ->
-                         With_hash.
-                           { data = Pickles.Side_loaded.Verification_key.dummy
-                           ; hash = Zkapp_account.dummy_vk_hash ()
-                           }
-                     | Some vk ->
-                         vk
-                   in
-                   Some
-                     { Zkapp_account.default with verification_key = Some vk }
-                  )
+                  Some
+                    { Zkapp_account.default with
+                      verification_key = Some verification_key
+                    }
               }
             else account_with_pk
           in
-          (* add new account to ledger *)
-          ( match
-              Ledger.get_or_create_account ledger
-                (Account_id.create account.public_key account.token_id)
-                account
-            with
-          | Ok (`Added, _) ->
-              ()
-          | Ok (`Existed, _) ->
-              failwith "account for new party already in ledger"
-          | Error err ->
-              failwithf "could not add account to ledger for new party: %s"
-                (Error.to_string_hum err) () ) ;
-          account )
+          return account )
     else
       match account_id with
       | None ->
-          (* choose an account from the ledger *)
-          let%map index =
-            Int.gen_uniform_incl 0 (Ledger.num_accounts ledger - 1)
-          in
-          let account = Ledger.get_at_index_exn ledger index in
-          (*get the latest state of this account*)
-          let (account : Account.t) =
-            Account_id.Table.find_exn account_state_tbl
-              (Account.identifier account)
+          if zkapp_account then
+            let%map zkapp_account_id =
+              Quickcheck.Generator.of_list zkapp_account_ids
+            in
+            match Account_id.Table.find account_state_tbl zkapp_account_id with
+            | None ->
+                failwith "gen_party_body: fail to find zkapp account"
+            | Some (_, `Fee_payer) | Some (_, `New_account) ->
+                failwith
+                  "gen_party_body: all zkapp accounts were new accounts or \
+                   used as fee_payer accounts"
+            | Some (acct, `Ordinary_participant) ->
+                acct
+          else
+            let accts =
+              Account_id.Table.filteri account_state_tbl
+                ~f:(fun ~key:_ ~data:(_, role) ->
+                  match role with
+                  | `Fee_payer | `New_account ->
```

Comment:
In the case of non-zkapp accounts, we could reuse newly created accounts, right?

Reply:
If we only stick to `signature` and `non-given` authorization.